### PR TITLE
add kubevela

### PR DESCRIPTION
**Pull request by Dentrax**
This PR adds kubevela (apiserver/manager/cli).

Signed-off-by: Furkan <furkan.turkal@trendyol.com>

### CODE REVIEW

**Comment by Dentrax** [comment]
@tuananh I was following `CONTRIBUTING.md` but somehow I couldn't able to get success build from _any_ package due to `ERROR: unable to select packages:` error. I'm on Mac. Am I missing something? Should I use lima instead `dev-container`?

**Comment by tuananh** [comment]
> @tuananh I was following `CONTRIBUTING.md` but somehow I couldn't able to get success build from _any_ package due to `ERROR: unable to select packages:` error. I'm on Mac. Am I missing something? Should I use lima instead `dev-container`?

can you post more error logs?

File: kubevela.yaml [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+      - busybox
+      - go
+      - build-base
+      - git
+pipeline:
+  - uses: fetch
+    with:
+      uri: https://github.com/kubevela/kubevela/archive/v${{package.version}}/v${{package.version}}.tar.gz
+      expected-sha256: 0b4d43518692e5c758148d1f8d78fe98f3134082cb4b3b49ca1200e3e44e8ade
+  - runs: |
```
Comment (developer-guy) [inline kubevela.yaml:28]: I found two targets in the makefile of the kubevela, which abstracts all the details for us and makes this process easy to read, WDYT? It seems that the only thing that we need to do is run the "make vela-cli and make manager" and then install the binaries to the target dir of melange:

➡️https://github.com/kubevela/kubevela/blob/master/Makefile#L25
➡️https://github.com/kubevela/kubevela/blob/master/makefiles/build.mk#L3

**Comment by developer-guy** [comment]
I'm using the lima approach:
➡️https://github.com/chainguard-dev/melange/tree/main/mac

```shell
melange build -r https://packages.wolfi.dev/os -k /usr/lib/wolfi-signing.rsa.pub --arch x86_64 kubevela.yaml
```

**Comment by Dentrax** [comment]
> > @tuananh I was following `CONTRIBUTING.md` but somehow I couldn't able to get success build from _any_ package due to `ERROR: unable to select packages:` error. I'm on Mac. Am I missing something? Should I use lima instead `dev-container`?
> 
> can you post more error logs?

Where is full logs:

<details>

  ```
  [sdk] ❯ make packages/kubevela
Makefile:166: warning: overriding recipe for target 'packages/x86_64/http-parser-2.9.4-r0.apk'
Makefile:164: warning: ignoring old recipe for target 'packages/x86_64/http-parser-2.9.4-r0.apk'
Makefile:243: warning: overriding recipe for target 'packages/x86_64/wasi-libc-0.20220525-r0.apk'
Makefile:169: warning: ignoring old recipe for target 'packages/x86_64/wasi-libc-0.20220525-r0.apk'
mkdir -p ./kubevela/
SOURCE_DATE_EPOCH=0 /usr/bin/melange build kubevela.yaml --repository-append /Users/furkan.turkal/src/public/os/packages --keyring-append local-melange.rsa.pub --signing-key local-melange.rsa --pipeline-dir /usr/share/melange/pipelines --arch x86_64 --env-file build-x86_64.env --cache-dir gs://wolfi-sources/ --namespace wolfi  --source-dir ./kubevela/
2023/01/13 09:20:14 building for [amd64]
2023/01/13 09:20:14 melange: detected git commit for build configuration: 68d6405081f4f7dcb7fd5491ed5d6cae3c4312af
2023/01/13 09:20:14 melange (kubevela/x86_64): melange is building:
2023/01/13 09:20:14 melange (kubevela/x86_64):   configuration file: kubevela.yaml
2023/01/13 09:20:14 melange (kubevela/x86_64):   workspace dir: /tmp/melange-workspace-3823748281
2023/01/13 09:20:14 melange (kubevela/x86_64): evaluating pipelines for package requirements
2023/01/13 09:20:14 melange (kubevela/x86_64):   adding package "wget" for pipeline "Fetch and extract external object into workspace"
2023/01/13 09:20:14 melange (kubevela/x86_64): building workspace in '/tmp/melange-guest-1246189302' with apko
Jan 13 09:20:14.096 [INFO] [arch:x86_64] build context:
Jan 13 09:20:14.096 [INFO] [arch:x86_64]   working directory: /tmp/melange-guest-1246189302
Jan 13 09:20:14.096 [INFO] [arch:x86_64]   tarball path:
Jan 13 09:20:14.096 [INFO] [arch:x86_64]   use proot: false
Jan 13 09:20:14.096 [INFO] [arch:x86_64]   source date: 1970-01-01 00:00:00 +0000 UTC
Jan 13 09:20:14.096 [INFO] [arch:x86_64]   Docker mediatypes: false
Jan 13 09:20:14.096 [INFO] [arch:x86_64]   SBOM output path:
Jan 13 09:20:14.096 [INFO] [arch:x86_64]   arch: x86_64
Jan 13 09:20:14.096 [INFO] [arch:x86_64] image configuration:
Jan 13 09:20:14.096 [INFO] [arch:x86_64]   contents:
Jan 13 09:20:14.096 [INFO] [arch:x86_64]     repositories: []
Jan 13 09:20:14.096 [INFO] [arch:x86_64]     keyring:      []
Jan 13 09:20:14.097 [INFO] [arch:x86_64]     packages:     [build-base busybox ca-certificates-bundle git go wget]
Jan 13 09:20:14.097 [INFO] [arch:x86_64]   accounts:
Jan 13 09:20:14.097 [INFO] [arch:x86_64]     runas:
Jan 13 09:20:14.097 [INFO] [arch:x86_64]     users:
Jan 13 09:20:14.097 [INFO] [arch:x86_64]       - uid=1000(build) gid=1000
Jan 13 09:20:14.097 [INFO] [arch:x86_64]     groups:
Jan 13 09:20:14.097 [INFO] [arch:x86_64]       - gid=1000(build) members=[build]
Jan 13 09:20:14.097 [INFO] [arch:x86_64] doing pre-flight checks
Jan 13 09:20:14.097 [INFO] [arch:x86_64] building image fileystem in /tmp/melange-guest-1246189302
Jan 13 09:20:14.097 [INFO] [arch:x86_64] initializing apk database
Jan 13 09:20:14.097 [INFO] [arch:x86_64] [cmd:apk] [use-proot:false] [use-qemu:] running: /sbin/apk add --initdb --arch x86_64 --root /tmp/melange-guest-1246189302
Jan 13 09:20:14.136 [DEBUG] [arch:x86_64] [cmd:apk] [use-proot:false] [use-qemu:] OK: 0 MiB in 0 packages
Jan 13 09:20:14.140 [INFO] [arch:x86_64] initializing apk keyring
Jan 13 09:20:14.141 [INFO] [arch:x86_64] initializing apk repositories
Jan 13 09:20:14.140 [INFO] [arch:x86_64] initializing apk world
Jan 13 09:20:14.150 [DEBUG] [arch:x86_64] appending 1 extra keys to keyring
Jan 13 09:20:14.150 [DEBUG] [arch:x86_64] installing key local-melange.rsa.pub
Jan 13 09:20:14.150 [DEBUG] [arch:x86_64] installing key /usr/share/apk/keys/x86_64/alpine-devel@lists.alpinelinux.org-4a6a0840.rsa.pub
Jan 13 09:20:14.150 [DEBUG] [arch:x86_64] installing key /usr/share/apk/keys/x86_64/alpine-devel@lists.alpinelinux.org-6165ee59.rsa.pub
Jan 13 09:20:14.150 [DEBUG] [arch:x86_64] installing key /usr/share/apk/keys/x86_64/alpine-devel@lists.alpinelinux.org-5261cecb.rsa.pub
Jan 13 09:20:14.152 [INFO] [arch:x86_64] synchronizing with desired apk world
Jan 13 09:20:14.152 [INFO] [arch:x86_64] [cmd:apk] [use-proot:false] [use-qemu:] running: /sbin/apk fix --root /tmp/melange-guest-1246189302 --no-scripts --no-cache --update-cache --arch x86_64
Jan 13 09:20:14.158 [DEBUG] [arch:x86_64] [cmd:apk] [use-proot:false] [use-qemu:] WARNING: Ignoring /Users/furkan.turkal/src/public/os/packages: No such file or directory
Jan 13 09:20:14.158 [DEBUG] [arch:x86_64] [cmd:apk] [use-proot:false] [use-qemu:] ERROR: unable to select packages:
Jan 13 09:20:14.159 [DEBUG] [arch:x86_64] [cmd:apk] [use-proot:false] [use-qemu:]   build-base (no such package):
Jan 13 09:20:14.159 [DEBUG] [arch:x86_64] [cmd:apk] [use-proot:false] [use-qemu:]     required by: world[build-base]
Jan 13 09:20:14.159 [DEBUG] [arch:x86_64] [cmd:apk] [use-proot:false] [use-qemu:]   busybox (no such package):
Jan 13 09:20:14.159 [DEBUG] [arch:x86_64] [cmd:apk] [use-proot:false] [use-qemu:]     required by: world[busybox]
Jan 13 09:20:14.159 [DEBUG] [arch:x86_64] [cmd:apk] [use-proot:false] [use-qemu:]   ca-certificates-bundle (no such package):
Jan 13 09:20:14.159 [DEBUG] [arch:x86_64] [cmd:apk] [use-proot:false] [use-qemu:]     required by: world[ca-certificates-bundle]
Jan 13 09:20:14.160 [DEBUG] [arch:x86_64] [cmd:apk] [use-proot:false] [use-qemu:]   git (no such package):
Jan 13 09:20:14.160 [DEBUG] [arch:x86_64] [cmd:apk] [use-proot:false] [use-qemu:]     required by: world[git]
Jan 13 09:20:14.160 [DEBUG] [arch:x86_64] [cmd:apk] [use-proot:false] [use-qemu:]   go (no such package):
Jan 13 09:20:14.160 [DEBUG] [arch:x86_64] [cmd:apk] [use-proot:false] [use-qemu:]     required by: world[go]
Jan 13 09:20:14.160 [DEBUG] [arch:x86_64] [cmd:apk] [use-proot:false] [use-qemu:]   wget (no such package):
Jan 13 09:20:14.160 [DEBUG] [arch:x86_64] [cmd:apk] [use-proot:false] [use-qemu:]     required by: world[wget]
2023/01/13 09:20:14 ERROR: failed to build package. the build environment has been preserved:
2023/01/13 09:20:14 melange (kubevela/x86_64):   workspace dir: /tmp/melange-workspace-3823748281
2023/01/13 09:20:14 melange (kubevela/x86_64):   guest dir: /tmp/melange-guest-1246189302
Error: failed to build package: unable to build guest: unable to generate image: initializing apk: failed to fixate apk world: exit status 6
2023/01/13 09:20:14 error during command execution: failed to build package: unable to build guest: unable to generate image: initializing apk: failed to fixate apk world: exit status 6
make: *** [Makefile:353: packages/x86_64/kubevela-1.6.7-r0.apk] Error 1
  ```
</details>

File: kubevela.yaml (hunk 1 above)
Comment (Dentrax) [reply kubevela.yaml:28]: Thanks, have already noticed. But some packages seem used non-make way (even though projects provide a Makefile), so I'm curious if it's intentional. I can change to `make` anyway. 🤔

File: kubevela.yaml [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+      GOLDFLAGS="-s -w
+        -X github.com/oam-dev/kubevela/version.VelaVersion=${{package.version}}
+        -X github.com/oam-dev/kubevela/version.GitRevision=git-$(git rev-parse --short HEAD)
```
Comment (tuananh) [inline kubevela.yaml:31]: this shouldn't work because the tarball you download is not a git directory

Comment (Dentrax) [reply kubevela.yaml:31]: Very good point! Any advices for this case? Maybe I could set: `github.com/oam-dev/kubevela/version.GitRevision=Wolfi` etc.

Comment (tuananh) [reply kubevela.yaml:31]: @Dentrax i don't know about this either. maybe @rawlingsj can advise?

**Comment by tuananh** [comment]
@Dentrax btw, there's `go/build` pipeline built-in in melange

it looks like this https://gist.github.com/tuananh/7bf4996b6c2d4d6775ddd1d4d6b09fca

also, i tested the dev-container and it's working for me, using the above yaml file :-/

image

File: kubevela.yaml (hunk 2 above)
Comment (Dentrax) [reply kubevela.yaml:31]: I'm thinking if its worth to add a lint check for this case: _"given download is tarball and `$(git` shell command is contains in `runs` field."_ One pitfall of this is that we couldn't catch if `make` is used, and it calls `git` command under the hood.

**Comment by Dentrax** [comment]
> @Dentrax btw, there's `go/build` pipeline built-in in melange
> 
> it looks like this [gist.github.com/tuananh/7bf4996b6c2d4d6775ddd1d4d6b09fca](https://gist.github.com/tuananh/7bf4996b6c2d4d6775ddd1d4d6b09fca)

Oh, your approach is much better and clear IMO. You duplicated `ldflags` because melange does not support _inheritance_ in _pipeline_, right? Maybe I should file a proposal for this. 🤔

If you are OK with this, I can directly use yours!

> also, i tested the dev-container and it's working for me

I think it's because my macOS environment, somehow melange couldn't install necessary packages and eventually throws _not found_ errors.

**Comment by tuananh** [comment]
> You duplicated ldflags because melange does not support inheritance in pipeline, right?

Not sure, I haven't tried it yet.

> If you are OK with this, I can directly use yours!

Yes, pls. Though it's still missing the git commit ldflags above so the compiled binary will not have that gitcommit info.

**Comment by Dentrax** [comment]
> Yes go ahead. Though it's still missing the git commit ldflags above so the compiled binary will not have that gitcommit info.

Thanks, updated & added you as co-author.

File: Makefile [hunk 3]
```diff
@@ -349,6 +349,7 @@ $(eval $(call build-package,py3-s3transfer,0.5.1-r0))
 $(eval $(call build-package,py3-urllib3,1.26.13-r0))
 $(eval $(call build-package,py3-yaml,6.0-r0))
 $(eval $(call build-package,aws-cli,1.27.38-r0))
+$(eval $(call build-package,kubevela,1.6.7-r0))
```
Comment (developer-guy) [inline Makefile:352]: One small thing here that causes pipeline to crash, you are using `1.6.7` version here but the actual package version is `1.7.0`, you need to upgrade it in here too.

Comment (Dentrax) [reply Makefile:352]: Good catch! I think we should add a check to linter for this case.

Comment (developer-guy) [reply Makefile:352]: There is already a check for this in lint.sh where I found this error.

**Review by developer-guy — APPROVED** [review]
LGTM

File: kubevela.yaml [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+      - go
+      - build-base
+pipeline:
+  - uses: fetch
+    with:
+      uri: https://github.com/kubevela/kubevela/archive/refs/tags/v${{package.version}}.tar.gz
+      expected-sha256: fd102cc6fd76273d5a3bc88faa050f3d0eed5cbfc9a253522d9fad012cc0babf
+  - runs: |
+      export CGO_ENABLED=0
+      export GO111MODULE=on
```
Comment (jdolitsky) [inline kubevela.yaml:29]: do these exports persist to the next step?

**Review by jdolitsky — APPROVED** [review]

File: kubevela.yaml (hunk 4 above)
Comment (Dentrax) [reply kubevela.yaml:29]: any ideas @tuananh?

Comment (developer-guy) [reply kubevela.yaml:29]: AFAIK, they don't ☝️

**Review by dlorenc — APPROVED** [review]

**Pull request merged**